lib/src/cli/flutter_cli.dart

```diff
@@ -65,55 +62,182 @@ class Flutter {
   static Future<void> test({
     String cwd = '.',
     bool recursive = false,
-    void Function([String?]) Function(String message)? progress,
-  }) async {
-    await _runCommand(
-      cmd: (cwd) async {
-        final installDone = progress?.call(
-          'Running "flutter test" in $cwd',
+    void Function(String)? stdout,
+    void Function(String)? stderr,
+  }) {
+    return _runCommand(
+      cmd: (cwd) {
+        void noop(String? _) {}
+        stdout?.call('Running "flutter test" in $cwd...\n');
+        return _flutterTest(
+          cwd: cwd,
+          stdout: stdout ?? noop,
+          stderr: stderr ?? noop,
         );
-        try {
-          final result = await _Cmd.run(
-            'flutter',
-            ['test'],
-            workingDirectory: cwd,
-          );
-          return result;
-        } catch (_) {
-          rethrow;
-        } finally {
-          installDone?.call();
-        }
       },
       cwd: cwd,
       recursive: recursive,
     );
   }
+}
+
+/// Run a command on directories with a `pubspec.yaml`.
+Future<void> _runCommand<T>({
+  required Future<T> Function(String cwd) cmd,
+  required String cwd,
+  required bool recursive,
+}) async {
+  if (!recursive) {
+    final pubspec = File(p.join(cwd, 'pubspec.yaml'));
+    if (!pubspec.existsSync()) throw PubspecNotFound();
+
+    await cmd(cwd);
+    return;
+  }
 
-  /// Run a command on directories with a `pubspec.yaml`.
-  static Future<void> _runCommand({
-    required Future<ProcessResult> Function(String cwd) cmd,
-    required String cwd,
-    required bool recursive,
-  }) async {
-    if (!recursive) {
-      final pubspec = File(p.join(cwd, 'pubspec.yaml'));
-      if (!pubspec.existsSync()) throw PubspecNotFound();
+  final processes = _Cmd.runWhere(
+    run: (entity) => cmd(entity.parent.path),
+    where: _isPubspec,
+    cwd: cwd,
+  );
 
-      await cmd(cwd);
-      return;
-    }
+  if (processes.isEmpty) throw PubspecNotFound();
 
-    final processes = _Cmd.runWhere(
-      run: (entity) => cmd(entity.parent.path),
-      where: _isPubspec,
-      cwd: cwd,
-    );
+  for (final process in processes) {
```

Comment:
Do we need they to be in sequence? If not we could gain time by running in parallel?

Reply:
They do unfortunately, otherwise flutter processes with lock/conflict with each other. This is an optimization we can revisit in a separate PR but I believe is out of scope for this work.